### PR TITLE
Knowledge backend

### PR DESCRIPTION
**Pull request by Tarraann**
### Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs update

### Checklist
- [ ] My pull request is atomic and focuses on a single change.
- [x] I have read the contributing guide and my code conforms to the guidelines.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have added the required tests.

### CODE REVIEW

File: migrations/versions/111f4f3af57b_create_vector_db_index_table.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+
+# revision identifiers, used by Alembic.
+revision = '111f4f3af57b'
+down_revision = 'd3103b0ecef1'
+branch_labels = None
+depends_on = None
+
+
+def upgrade() -> None:
```
Comment (TransformerOptimus) [inline migrations/versions/111f4f3af57b_create_vector_db_index_table.py:19]: why do u need this migration?

File: migrations/versions/18e8e1ee3f03_create_knowledge_table.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+
+# revision identifiers, used by Alembic.
+revision = '18e8e1ee3f03'
+down_revision = '49a07c84feee'
+branch_labels = None
+depends_on = None
+
+
+def upgrade() -> None:
```
Comment (TransformerOptimus) [inline migrations/versions/18e8e1ee3f03_create_knowledge_table.py:19]: why this migration?

File: migrations/versions/49a07c84feee_create_vector_db_table.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    sa.Column('summary', sa.String(), nullable=True),
+    sa.Column('readme', sa.String(), nullable=True),
+    sa.Column('index_id', sa.Integer(), nullable=True),
+    sa.Column('organisation_id', sa.Integer(), nullable=True),
+    sa.Column('contributed_by', sa.String(), nullable=True),
+    sa.Column('created_at', sa.DateTime(), nullable=True),
+    sa.Column('updated_at', sa.DateTime(), nullable=True),
+    sa.PrimaryKeyConstraint('id')
+    )
+    op.create_table('marketplace_state',
```
Comment (TransformerOptimus) [inline migrations/versions/49a07c84feee_create_vector_db_table.py:34]: dont even create this table if u are dropping later.

File: migrations/versions/49a07c84feee_create_vector_db_table.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+# revision identifiers, used by Alembic.
+revision = '49a07c84feee'
+down_revision = 'ba60b12ae109'
+branch_labels = None
+depends_on = None
+
+
+def upgrade() -> None:
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_table('knowledge',
```
Comment (TransformerOptimus) [inline migrations/versions/49a07c84feee_create_vector_db_table.py:21]: all tables names should be plural.

File: migrations/versions/9d8fa6a0bd90_create_knowledge_config_table.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+# revision identifiers, used by Alembic.
+revision = '9d8fa6a0bd90'
+down_revision = '18e8e1ee3f03'
+branch_labels = None
+depends_on = None
+
+
+def upgrade() -> None:
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_table('knowledge_config',
```
Comment (TransformerOptimus) [inline migrations/versions/9d8fa6a0bd90_create_knowledge_config_table.py:21]: make it knowledge_configs

File: migrations/versions/9d8fa6a0bd90_create_knowledge_config_table.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    sa.Column('id', sa.Integer(), autoincrement=True, nullable=False),
+    sa.Column('knowledge_id', sa.Integer(), nullable=True),
```
Comment (TransformerOptimus) [inline migrations/versions/9d8fa6a0bd90_create_knowledge_config_table.py:23]: id, knowledge_id, key cant be nullable.

File: migrations/versions/49a07c84feee_create_vector_db_table.py (hunk 3 above)
Comment (TransformerOptimus) [reply migrations/versions/49a07c84feee_create_vector_db_table.py:34]: table name should be knowledges, vector_dbs, vector_db_index_collections

File: migrations/versions/49a07c84feee_create_vector_db_table.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    op.create_table('vector_db',
+    sa.Column('id', sa.Integer(), autoincrement=True, nullable=False),
+    sa.Column('name', sa.String(), nullable=True),
+    sa.Column('db_type', sa.String(), nullable=True),
+    sa.Column('organisation_id', sa.Integer(), nullable=True),
+    sa.Column('created_at', sa.DateTime(), nullable=True),
+    sa.Column('updated_at', sa.DateTime(), nullable=True),
+    sa.PrimaryKeyConstraint('id')
+    )
+    op.create_table('vector_db_index_collection',
```
Comment (TransformerOptimus) [inline migrations/versions/49a07c84feee_create_vector_db_table.py:53]: should be vector_db_indexes. index_collection does not make sense.

name, vector_db_id can't be nullable

File: superagi/controllers/knowledge.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+from superagi.helper.pinecone_helper import PineconeHelper
+from superagi.helper.qdrant_helper import QdrantHelper
+from superagi.helper.knowledge_helper import KnowledgeHelper
+from superagi.models.vector_db_config import VectordbConfig
+from superagi.helper.auth import get_user_organisation
+
+router = APIRouter()
+
+
+@router.get("/get/list")
```
Comment (TransformerOptimus) [inline superagi/controllers/knowledge.py:22]: marketplace/list

File: superagi/controllers/knowledge.py [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+    marketplace_knowledges_with_install = Knowledge.get_knowledge_installed_details(db.session, marketplace_knowledges,
+                                                                              organisation)
+    for knowledge in marketplace_knowledges_with_install:
+        knowledge["install_number"] = MarketPlaceStats.get_knowledge_installation_number(db.session, knowledge["id"])
+    return marketplace_knowledges_with_install
+
+@router.get("/user/list")
+def get_user_knowledge_list(Authorize: AuthJWT = Depends(), organisation = Depends(get_user_organisation)):
+    organisation_id = organisation.id
+    marketplace_organisation_id = int(get_config("MARKETPLACE_ORGANISATION_ID"))
```
Comment (TransformerOptimus) [inline superagi/controllers/knowledge.py:49]: logic does not seem right. local one cant even fetch MARKETPLACE_ORGANISATION_ID. In local nobody sets MARKETPLACE_ORGANISATION_ID

File: superagi/controllers/vector_db.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+from superagi.models.vector_db import Vectordb
+from superagi.models.vector_db_config import VectordbConfig
+from superagi.models.vector_db_index_collection import VectorIndexCollection
+from superagi.models.index_config import VectorIndexConfig
+from superagi.helper.pinecone_helper import PineconeHelper
+from superagi.helper.qdrant_helper import QdrantHelper
+from superagi.helper.auth import get_user_organisation
+router = APIRouter()
+@router.get("/get/list")
+def handle_marketplace_operations_list():
```
Comment (TransformerOptimus) [inline superagi/controllers/vector_db.py:14]: rename the method.

File: superagi/controllers/vector_db_index.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    
+    return vector_indices_data
+
+@router.get("/get/user/valid_indices")
+def get_user_valid_indices(organisation = Depends(get_user_organisation)):
+    vector_dbs = Vectordb.get_vector_db_organisation(db.session, organisation)
+    pinecone = []
+    qdrant = []
+    for vector in vector_dbs:
+        indices = db.session.query(VectorIndexCollection).filter(VectorIndexCollection.vector_db_id == vector["id"]).all()
```
Comment (TransformerOptimus) [inline superagi/controllers/vector_db_index.py:46]: not a right pattern to call sql inside a loop

File: superagi/controllers/vector_db_index.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+router = APIRouter()
+
+@router.get("/get/marketplace/valid_indices/{knowledge_id}")
+def get_marketplace_valid_indices(knowledge_id: int, organisation = Depends(get_user_organisation)):
+    vector_dbs = Vectordb.get_vector_db_organisation(db.session, organisation)
+    pinecone = []
+    qdrant = []
+    for vector in vector_dbs:
+        indices = db.session.query(VectorIndexCollection).filter(VectorIndexCollection.vector_db_id == vector["id"]).all()
+        for index in indices:
```
Comment (TransformerOptimus) [inline superagi/controllers/vector_db_index.py:20]: not a right pattern to call sql inside a loop

File: superagi/helper/knowledge_helper.py [hunk 13]
```diff
@@ -0,0 +1,35 @@
+from sqlalchemy.orm import Session
+import boto3
+import json
+from superagi.config.config import get_config
+from superagi.models.knowledge_config import KnowledgeConfig
+from superagi.models.index_config import VectorIndexConfig
+
+class KnowledgeHelper:
```
Comment (TransformerOptimus) [inline superagi/helper/knowledge_helper.py:8]: helper classes does not make sense. Create a new package knowledge and restructure the classes.

File: migrations/versions/024c5c121f57_create_index_config_table.py [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+# revision identifiers, used by Alembic.
+revision = '024c5c121f57'
+down_revision = '111f4f3af57b'
+branch_labels = None
+depends_on = None
+
+
+def upgrade() -> None:
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_table('vector_index_config',
```
Comment (TransformerOptimus) [inline migrations/versions/024c5c121f57_create_index_config_table.py:21]: rename to vector_index_configs.

File: superagi/controllers/vector_db.py [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+        if index.name not in new_indices:
+            VectorIndexConfig.delete_vector_index_config(db.session, vector_index_id=index.id)
+            VectorIndexCollection.delete_vector_index(db.session, id=index.id)
+        existing_index_names.append(index.name)
+    existing_indices_names = set(existing_index_names)
+    new_indices_names = set(new_indices)
+    added_indices = new_indices_names - existing_indices_names
+    for index in added_indices:
+        vector_index = VectorIndexCollection.add_vector_index(db.session, existing_index, vector_db_id)
+        if vector_db.db_type == "Pinecone":
```
Comment (TransformerOptimus) [inline superagi/controllers/vector_db.py:88]: We cant have if else for each db type can we have better structure.

File: superagi/helper/qdrant_helper.py [hunk 16]
```diff
@@ -0,0 +1,117 @@
+from sqlalchemy.orm import Session
+from superagi.models.vector_db_config import VectordbConfig
+from qdrant_client import models, QdrantClient
+
+class QdrantHelper:
```
Comment (TransformerOptimus) [inline superagi/helper/qdrant_helper.py:5]: We already have these vector dbs in the vector_dbs. These classes does not make sense.

File: superagi/models/index_config.py [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+    
+    @classmethod
+    def add_vector_index_config(cls, session, vector_index_id, key, value):
+        index_config = VectorIndexConfig(vector_index_id=vector_index_id, key=key, value=value)
+        session.add(index_config)
+        session.commit()
+
+    @classmethod
+    def delete_vector_index_config(cls, session, vector_index_id):
+        session.query(VectorIndexConfig).filter(VectorIndexConfig.vector_index_id == vector_index_id).delete()
```
Comment (TransformerOptimus) [inline superagi/models/index_config.py:42]: if no element found delete might throw exception

File: superagi/tools/knowledge_tool/knowledge.py [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+        session = self.toolkit_config.session
+        knowledge_id = session.query(AgentConfiguration).filter(
+            AgentConfiguration.agent_id == self.agent_id,
+            AgentConfiguration.key == "knowledge").first()
+        knowledge = session.query(Knowledge).filter(Knowledge.id == knowledge_id).first()
+        dbhelper = KnowledgeToolDbHelper(session)
+        knowledge_details = dbhelper.get_knowledge_details(knowledge)
+        query_knowledge = KnowledgeToolHelper()
+
+        if knowledge_details["knowledge_vector_db_type"] == "Pinecone":
```
Comment (TransformerOptimus) [inline superagi/tools/knowledge_tool/knowledge.py:46]: Get a better way than if else

File: superagi/models/knowledge.py [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+    """
+
+    __tablename__ = 'knowledge'
+
+    id = Column(Integer, primary_key=True, autoincrement=True)
+    name = Column(String)
+    description = Column(String)
+    summary = Column(String)
+    readme = Column(String)
+    index_id = Column(Integer)
```
Comment (TransformerOptimus) [inline superagi/models/knowledge.py:34]: vector_db_index_id

File: superagi/models/index_config.py [hunk 20]
```diff
@@ -0,0 +1,48 @@
+from sqlalchemy import Column, Integer, Text, String
+
+from superagi.models.base_model import DBBaseModel
+
+
+class VectorIndexConfig(DBBaseModel):
```
Comment (TransformerOptimus) [inline superagi/models/index_config.py:6]: clean up the file_name.

Comment (TransformerOptimus) [reply superagi/models/index_config.py:6]: vector_db_index_configs

**Review by DMTarmey — COMMENTED** [review]
cheers

**Review by DMTarmey — COMMENTED** [review]
cheers

**Pull request closed without merging**